test/server/lib/DocApi.ts

```diff
@@ -4462,6 +4475,161 @@ function testDocApi() {
+    assert.equal(resp.status, 400);
+    assert.deepEqual(resp.data, {
+      error: 'Invalid payload',
+      details: { userError: 'Error: value.sql is missing' }
+    });
+  });
+
+  it ("POST /docs/{did}/sql has access control", async function () {
+    // Check non-viewer doesn't have access.
+    const url = `${homeUrl}/api/docs/${docIds.Timesheets}/sql`;
+    const query = { sql: "select A from Table1 where id = ?", args: [ 1 ] };
+    let resp = await axios.post(url, query, kiwi);
+    assert.equal(resp.status, 403);
+    assert.deepEqual(resp.data, {
+      error: 'No view access',
+    });
+
+    try {
+      // Check a viewer would have access.
+      const delta = {
+        users: { 'kiwi@getgrist.com': 'viewers' },
+      };
+      await axios.patch(`${homeUrl}/api/docs/${docIds.Timesheets}/access`, {delta}, chimpy);
+      await flushAuth();
+      resp = await axios.get(`${homeUrl}/api/docs/${docIds.Timesheets}/access`, chimpy);
+      resp = await axios.post(url, query, kiwi);
+      assert.equal(resp.status, 200);
+
+      // Check a viewer would not have access if there is some private material.
+      await axios.post(
+        `${homeUrl}/api/docs/${docIds.Timesheets}/apply`, [
+          ['AddTable', 'TablePrivate', [{id: 'A', type: 'Int'}]],
+          ['AddRecord', '_grist_ACLResources', -1, {tableId: 'TablePrivate', colIds: '*'}],
+          ['AddRecord', '_grist_ACLRules', null, {
+            resource: -1, aclFormula: '', permissionsText: 'none',
+          }],
+        ], chimpy);
+      resp = await axios.post(url, query, kiwi);
+      assert.equal(resp.status, 403);
+    } finally {
+      // Remove extra viewer; remove extra table.
+      const delta = {
+        users: { 'kiwi@getgrist.com': null },
+      };
+      await axios.patch(`${homeUrl}/api/docs/${docIds.Timesheets}/access`, {delta}, chimpy);
+      await flushAuth();
+      await axios.post(
+        `${homeUrl}/api/docs/${docIds.Timesheets}/apply`, [
+          ['RemoveTable', 'TablePrivate'],
+        ], chimpy);
+    }
+  });
+
+  it ("POST /docs/{did}/sql accepts only selects", async function () {
+    async function check(accept: boolean, sql: string, ...args: any[]) {
+      const resp = await axios.post(
+        `${homeUrl}/api/docs/${docIds.Timesheets}/sql`,
+        { sql, args },
+        chimpy);
+      if (accept) {
+        assert.equal(resp.status, 200);
+      } else {
+        assert.equal(resp.status, 400);
+      }
+      return resp.data;
+    }
+    await check(true, 'select * from Table1');
+    await check(true, '  SeLeCT * from Table1');
+    await check(true, 'with results as (select 1) select * from results');
+
+    // rejected quickly since no select
+    await check(false, 'delete from Table1');
+
+    // rejected because deletes/updates/... can't be nested within a select
+    await check(false, 'delete from Table1 where id in (select id from Table1)');
+    await check(false, 'update Table1 set A = ?', 'test');
```

Comment:
I think there are more checks to add in order to be safe against SQLi.

If I add this test, it seems like it does not pass the check (i.e. the request is accepted):
```js
    await check(false, 'select id from Table1); delete from Table1; (select id from Table1');
```

Reply:
That is similar in nature to this check a little later, right?
```
await check(true, 'select * from Table1); delete from Table1 where id in (select id from Table1');
```

My comment there was:
```
   // Of course, we can get out of the wrapping select, but we can't
    // add on more statements. For example, the following runs with no
    // trouble - but only the SELECT part. The DELETE is discarded.
    // (node-sqlite3 doesn't expose enough to give an error message for
    // this, though we could extend it).
```

So: security wise, only the first statement goes through. It would be great to have an error, but I don't see a way to do it without modifying node-sqlite3 (which we've done before, it is painful though) or adding a parser for SQLite flavor of SQL (which, from a security perspective, we couldn't trust anyway, since the parser might be missing a case). Adding a parser would allow giving a better error message for users who try statements like this, so I'm open to it.

Reply:
These two points in the PR description are also relevant:
 * The underlying SQLite functions used will only process the first statement in the supplied text (the remainder is placed in a "tail string" ignored by node-sqlite3). ...
 * The text is wrapped in `select * from (USER SUPPLIED TEXT)` ... It is straightforward to break out of such a wrapper with multiple statements, but again, only the first statement is processed.

Reply:
One thing I've been looking into is other sqlite wrappers. For example, in grist-static, we replace node-sqlite3 with sql.js. Reading its API docs carefully, the call to `all` should be safe in the same way, but it is a bit scary how carefully one needs to read. That said, specifically in grist-static, the endpoint wouldn't even be available, and would only be accessing something in the user's own browser page.

Reply:
If you are still a bit worrying about this issue, do you think we may make a temporary copy of the SQLite file in order to query it?
The original file would remain untouched, so if an attacker achieved to alter the document, it won't have any effect.

One drawback I see is that it's more expansive with the IO though.

Reply:
I hadn't thought of a copy. That's a thought. Another option is a temporary read-only connection to the file (temporary to avoid the possibility of changes to the connection via pragmas). I didn't jump into doing that, since as soon as there is more than one connection to the sqlite file, file-level locking can kick in and some new logic would be needed to handle busy timeouts. Definitely doable though.

Reply:
Crazy additional idea - could offer PostgreSQL-compatible access via [postlite](https://github.com/benbjohnson/postlite)